src/pyqasm/visitor.py

```diff
@@ -1331,44 +1331,83 @@ def _visit_branching_statement(
+
+            if_block = self.visit_basic_block(statement.if_block)
+            else_block = self.visit_basic_block(statement.else_block)
+
+            if reg_idx is not None:
+                # single bit branch
                 Qasm3Validator.validate_register_index(
-                    reg_id, self._global_creg_size_map[reg_name], qubit=False
+                    reg_idx, self._global_creg_size_map[reg_name], qubit=False
                 )
 
-            new_lhs = (
-                qasm3_ast.IndexExpression(
-                    collection=qasm3_ast.Identifier(name=reg_name),
-                    index=[qasm3_ast.IntegerLiteral(reg_id)],
+                new_if_block = qasm3_ast.BranchingStatement(
+                    condition=qasm3_ast.BinaryExpression(
+                        op=qasm3_ast.BinaryOperator["=="],
+                        lhs=qasm3_ast.IndexExpression(
+                            collection=qasm3_ast.Identifier(name=reg_name),
+                            index=[qasm3_ast.IntegerLiteral(reg_idx)],
+                        ),
+                        rhs=(
+                            qasm3_ast.BooleanLiteral(rhs_value)
+                            if isinstance(rhs_value, bool)
+                            else qasm3_ast.IntegerLiteral(rhs_value)
+                        ),
+                    ),
+                    if_block=if_block,
+                    else_block=else_block,
                 )
-                if reg_id is not None
-                else qasm3_ast.Identifier(name=reg_name)
-            )
-            assert isinstance(rhs_value, (bool, int))
-            new_rhs = (
-                qasm3_ast.BooleanLiteral(rhs_value)
-                if isinstance(rhs_value, bool)
-                else qasm3_ast.IntegerLiteral(rhs_value)
-            )
-
-            new_if_block = qasm3_ast.BranchingStatement(
-                condition=qasm3_ast.BinaryExpression(
-                    op=qasm3_ast.BinaryOperator["=="],
-                    lhs=new_lhs,
-                    rhs=new_rhs,
-                ),
-                if_block=self.visit_basic_block(statement.if_block),
-                else_block=self.visit_basic_block(statement.else_block),
-            )
-            result.append(new_if_block)
+                result.append(new_if_block)
+            else:
+                # unroll multi-bit branch
+                assert isinstance(rhs_value, int) and op in [
+                    qasm3_ast.BinaryOperator[o] for o in ["==", ">=", "<=", ">", "<"]
+                ]
 
+                if op == qasm3_ast.BinaryOperator[">"]:
+                    op = qasm3_ast.BinaryOperator[">="]
+                    rhs_value += 1
+                elif op == qasm3_ast.BinaryOperator["<"]:
+                    op = qasm3_ast.BinaryOperator["<="]
+                    rhs_value -= 1
+
+                size = self._global_creg_size_map[reg_name]
+                rhs_value_str = bin(int(rhs_value))[2:].zfill(size)
```

Comment:
Can you double check if this works for negative values as well? Will be nice to have a test for negative values

Reply:
How can we support negatives at all if LHS is bit[]?

Reply:
Well negatives can be implemented using something like a sign bit or the two's complement method. So we can compare against that representation for any negative number. 

Will have to see the repr in Python and accordingly deal with it. Note that this reserves 1 bit for the sign.

Reply:
Yes, but do we want to restrict the representation like this? This means that bit[4] encodes -7 to 8 as opposed to 0 to 15. Adding this would make users have to rememeber to keep the sign bit, which doesn't seem conventional. I feel like most quantum algorithms with classical measurement record unsigned values?

Reply:
Yea that's a valid point, let's keep the logic for non-negative numbers only